src/hotspot/share/runtime/lockStack.inline.hpp

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_RUNTIME_LOCKSTACK_INLINE_HPP
+#define SHARE_RUNTIME_LOCKSTACK_INLINE_HPP
+
+#include "memory/iterator.hpp"
+#include "runtime/javaThread.hpp"
+#include "runtime/lockStack.hpp"
+
+inline int LockStack::to_index(int offset) {
+  return (offset - in_bytes(JavaThread::lock_stack_base_offset())) / oopSize;
+}
+
+inline bool LockStack::can_push() const {
+  return to_index(_offset) < CAPACITY;
+}
+
+inline void LockStack::push(oop o) {
+  validate("pre-push");
+  assert(oopDesc::is_oop(o), "must be");
+  assert(!contains(o), "entries must be unique");
+  assert(can_push(), "must have room");
+  _base[to_index(_offset)] = o;
+  _offset += oopSize;
+  validate("post-push");
+}
+
+inline oop LockStack::pop() {
+  validate("pre-pop");
+  assert(to_index(_offset) > 0, "underflow, probably unbalanced push/pop");
+  _offset -= oopSize;
+  oop o = _base[to_index(_offset)];
```

Comment:
I think `pop` and `remove` might benefit from zapping the removed elements in the stack, just to make sure we don't accidentally read them? `validate` can check that everything beoynd `end` is zapped, and there are no zapped elements on stack?

Reply:
I agree. We need to do this in Macroassembler::fast_unlock too, then.